### ISSUE TITLE
evaluate model alignment

### ISSUE THREAD

**Issue by McHaillet** [warpem/miss-alignment#22]

**Comment by McHaillet** [comment]
Alignment of the model with initial training runs makes the reconstruction (visually) worse on the evaluation tilt-series. After the meeting this wednesday I also tested it on the training tilt-series where it also makes the alignment worse. At least it's not a problem of overfitting on the training data.

**Comment by alisterburt** [comment]
@McHaillet I thought I had responded here - this obviously kinda sucks, given that you're still experimenting with the CNN it's not too discouraging though. I think the path forwards here is to get the CNN training working a little better and hopefully the better CNN will be better at driving alignments, what do you think?

**Comment by McHaillet** [comment]
Thought I should log back here a bit for documentation purposes.
* the bootstrapping approach is promising, this means training the model, aligning the tilt-series, warming-up and further retraining the model, realigning, etc...
* am still doing some bugfixing in getting stable training and automated early stopping with this approach, progress is good though

Below plots visualize the mean absolute difference of the y and x shifts to the ground truth alignments per iteration. The boxplot represents the performance over the 10 tilt-series:

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/17b09448-fa20-4411-9d15-7bcd067ac45a" />

<img width="500" alt="Image" src="https://github.com/user-attachments/assets/d95a44d1-0b11-49dc-a972-9c422f5ee3c2" />

AreTomo's (v1.3.4!) performance is on the right. I didn't use later AreTomo2 or AreTomo3 because I though version 1 would be fine given that we know the exact tomogram thickness (and dont need all the boiler plate). However, I cannot fix the tilt-axis-angle to 0 degrees, and for some tilt-series AreTomo optimizes it to a value that is a couple of degrees off (4.5 for example) . Which makes the performance significantly worse. I think if we could fix it to 0 degrees the performance and comparison would be a bit more 'fair'.

**Comment by alisterburt** [comment]
this is awesome, thanks for sharing! Definitely annoying that you can't fix the tilt axis angle for AreTomo 

At a later stage we'll have to compare against AT3 anyway so it might be worth getting that up and running

Interesting that optimization led worse alignments from iteration 2->3 then was able to recover!

**Comment by McHaillet** [comment]
> At a later stage we'll have to compare against AT3 anyway so it might be worth getting that up and running

The annoying thing is that you always need to start from an .mdoc and individual tilt images (AFAIK), which is a hassle for the simulated data because I would need to spoof all that stuff.

**Comment by McHaillet** [comment]
> Interesting that optimization led worse alignments from iteration 2->3 then was able to recover!

It could be due to the hyperparameter settings needing some more tuning, but it is nice indeed that it trucks along

**Comment by alisterburt** [comment]
> The annoying thing is that you always need to start from an .mdoc and individual tilt images (AFAIK), which is a hassle for the simulated data because I would need to spoof all that stuff. 

Definitely annoying :/

**Comment by dtegunov** [comment]
Looks great! I don't remember having issues with fixed tilt axis angles in AT1. Can you copy the command here?

**Comment by alisterburt** [comment]
https://gensoft.pasteur.fr/docs/AreTomo/1.3.4/AreTomoManual_1.3.0_09292022.pdf says 

```
-TiltAxis initVal refineFlag

Tilt axis determination is a two-step processing in AreTomo. A single tilt axis is first calculated
followed by the determination of how tilt axis varies over the entire tilt range. The “initVal” lets users
enter their estimate and AreTomo refines the estimate in [-3°, 3°]. If users do not know, they can enter 0
instead and AreTomo will search an optimal value in full range. The “refineFlag” can be one of -1, 0, and
1. The -1 instructs AreTomo to the user input value without any refinement. The default setting is 0 and
1.
```

is this what you tried?

**Comment by McHaillet** [comment]
Yep, that is what I did: `-TiltAxis 0 -1`

Maybe it needs to be a float value, or the 0 as the initVal might interfer with this behaviour
> If users do not know, they can enter 0 instead and AreTomo will search an optimal value in full range.

**Comment by McHaillet** [comment]
`-TiltAxis 0.00 -1` does not work to deactivate refinement, but `-TiltAxis 0.01 -1` does work 😆

**Comment by alisterburt** [comment]
lol - love it

**Comment by dtegunov** [comment]
Probably a parser hard-coded to stop parsing after -1/0/1. Did it affect the alignment accuracy?

**Comment by McHaillet** [comment]
The alignment accuracy is very similar to the figures above

**Comment by dtegunov** [comment]
@McHaillet could you please post a few XZ slice comparisons here when you have time? As Alister mentioned, they'd be very satisfying to look at 😊 I'd also be curious if there is a visible difference between the individual bootstrapping iterations, or if the model can see misalignment beyond what a human could discern.

**Comment by McHaillet** [comment]
sure thing! And I'll add the updated graph of the shift difference to ground truth and boxplots below

## Ours (tilt-series 1 from shrec)

start (tiltxcorr)

<img width="1836" height="643" alt="Image" src="https://github.com/user-attachments/assets/b5708acb-54aa-4e8f-aa49-97156ded861e" />

iteration 1

<img width="1836" height="643" alt="Image" src="https://github.com/user-attachments/assets/fb4ea57d-35b9-4bb4-991b-5a780736726b" />

iteration 2

<img width="1836" height="643" alt="Image" src="https://github.com/user-attachments/assets/a9f0bef2-b4f9-42e4-8521-9cec0228f268" />

iteration 3

<img width="1836" height="643" alt="Image" src="https://github.com/user-attachments/assets/63b4094e-2a24-4cca-a523-7494b8526c77" />

iteration 4

<img width="1836" height="643" alt="Image" src="https://github.com/user-attachments/assets/5f5de8d6-984c-4ac0-a37f-9469deb32534" />

iteration 5

<img width="1836" height="643" alt="Image" src="https://github.com/user-attachments/assets/9965d482-b141-40a1-887c-d156b732274d" />

## Ground truth

<img width="1836" height="643" alt="Image" src="https://github.com/user-attachments/assets/d865ecf8-9a75-4176-a0b6-4b7480bc0746" />

**Comment by McHaillet** [comment]
Absolute difference to ground truth for x and y shifts per tilt image:

<img width="600" alt="Image" src="https://github.com/user-attachments/assets/72c822bd-57a6-4299-8527-ccb62c753dc1" />

You can see the center tilts being a bit off here, especially in x.

**Comment by alisterburt** [comment]
So pretty! Looking forward to seeing you squash out those last little errors - you were going to try augmenting for them right?

**Comment by dtegunov** [comment]
Beautiful, thank you for compiling the full sequence! 

A question has been bugging me the whole weekend: can we make sure the model isn't just learning to make the high-contrast beads as spherical as possible? How many are there in each tomogram? Would they be easy to erase? They might also be the reason AT's alignment is so bad.

**Comment by McHaillet** [comment]
> They might also be the reason AT's alignment is so bad.

Could you explain you're thought there? My intuition is that the beads should improve the AT alignment

I will try if I can find a quick way to erase them :) (fidder or etomo might work)

**Comment by McHaillet** [comment]
At a quick glance I dont see immediate impact, with the caveat that the fiducial removal was not 100% accurate.

**original, i.e. with gold beads:** starting alignment from tiltxcorr in blue, alignment after optimization in orange. y is distance of alignment to ground truth

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/f465690c-b220-4f6e-8cf9-9a6a3e8f26ff" />

**with gold erasing:** exact same model as above only with gold bead erasing

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/c9632e80-a177-4ebc-9288-fda6f201afc3" />

This is the gold removal (left original, right erased):

<img width="600"  alt="Image" src="https://github.com/user-attachments/assets/1e777bd1-6c95-440a-afaf-88841b1d86e0" />

As you can see the gold is not fully removed, but a bunch of signal has definitely been deleted.

**Comment by dtegunov** [comment]
> > They might also be the reason AT's alignment is so bad.
> 
> Could you explain you're thought there? My intuition is that the beads should improve the AT alignment
> 
> I will try if I can find a quick way to erase them :) (fidder or etomo might work)

Sparse beads are great for patch tracking, but with AT-like algorithms you get a featureless cylinder along the Z axis for every bead after you project the 0 deg tilt image. Then in the next tilt, beads need to align to projections of those cylinders, and that doesn't go well.

**Comment by dtegunov** [comment]
> <img alt="Image" width="600" src="https://private-user-images.githubusercontent.com/58044494/476619876-1e777bd1-6c95-440a-afaf-88841b1d86e0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ5NDEwMTMsIm5iZiI6MTc1NDk0MDcxMywicGF0aCI6Ii81ODA0NDQ5NC80NzY2MTk4NzYtMWU3NzdiZDEtNmM5NS00NDBhLWFmYWYtODg4NDFiMWQ4NmUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODExVDE5MzE1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4ZTExZmQzZjM2NjE2MTNkNzg4NDBiYTRkZTUzMGI5ODIzNGE3MjdlZmZlYTVmZjc4ZWJlYmYyYjgxNGNkY2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.otirGl5u6TS0yp-AwfKJEB4Z38cSG6Hf2tkcVNbucwo">
> As you can see the gold is not fully removed, but a bunch of signal has definitely been deleted.

Since you'll be evaluating local alignment accuracy anyway, maybe it's worth picking some locations completely devoid of gold beads? (or just the center in the tomogram you plotted)
